packages/modules/fulfillment/src/services/fulfillment-module-service.ts

```diff
@@ -1416,28 +1687,141 @@ export default class FulfillmentModuleService<
     const updatedShippingOptionRules =
       await this.shippingOptionRuleService_.update(data_, sharedContext)
 
+    buildShippingOptionRuleEvents({
+      action: CommonEvents.UPDATED,
+      shippingOptionRules: updatedShippingOptionRules.map((rule) => ({
+        id: rule.id,
+      })),
+      sharedContext,
+    })
+
     return Array.isArray(data)
       ? updatedShippingOptionRules
       : updatedShippingOptionRules[0]
   }
 
-  @InjectTransactionManager("baseRepository_")
+  @InjectManager("baseRepository_")
+  @EmitEvents()
   async updateFulfillment(
     id: string,
     data: FulfillmentTypes.UpdateFulfillmentDTO,
     @MedusaContext() sharedContext: Context = {}
   ): Promise<FulfillmentTypes.FulfillmentDTO> {
-    const fulfillment = await this.fulfillmentService_.update(
-      { id, ...data },
-      sharedContext
+    const fulfillment = await this.updateFulfillment_(id, data, sharedContext)
+
+    return await this.baseRepository_.serialize<FulfillmentTypes.FulfillmentDTO>(
+      fulfillment
     )
+  }
 
-    const serialized =
-      await this.baseRepository_.serialize<FulfillmentTypes.FulfillmentDTO>(
-        fulfillment
+  @InjectTransactionManager("baseRepository_")
+  protected async updateFulfillment_(
+    id: string,
+    data: FulfillmentTypes.UpdateFulfillmentDTO,
+    @MedusaContext() sharedContext: Context
+  ): Promise<TFulfillmentEntity> {
+    const existingFulfillment: TFulfillmentEntity =
+      await this.fulfillmentService_.retrieve(
+        id,
+        {
+          relations: ["items", "labels"],
+        },
+        sharedContext
       )
 
-    return Array.isArray(serialized) ? serialized[0] : serialized
+    const updatedLabelIds: string[] = []
+    let deletedLabelIds: string[] = []
+
+    const existingLabelIds = existingFulfillment.labels.map((label) => label.id)
+
+    /**
+     * @note
+     * Since the relation is a one to many, the deletion, update and creation of labels
+     * is handled b the orm. That means that we dont have to perform any manual deletions or update.
+     * For some reason we use to have upsert and replace handled manually but we could simplify all that just like
+     * we do below which will create the label, update some and delete the one that does not exists in the new data.
+     *
+     * There is a bit of logic as we need to reassign the data of those we want to keep
+     * and we also need to emit the events later on.
+     */
+    if (isDefined(data.labels) && isPresent(data.labels)) {
```

Comment:
cc @sradevski here the upsert with replace works smoothly and handled by the orm on oneToMany relations, which is what we wanted, the only thing I am doing here is that if the user only provides the id I reassign the existing data to it. and also few things to emit the events as I need to know what will happen. Do you have an idea why we havn't done that in the product?

Reply:
I don't get the question, do you mean why did we use upsertWithReplace and not write it out manually like this?

Reply:
I am not writting it manually here, I need some info to prepare the event later i.e knowing what is created, updated, deleted but thats pretty much it. Here I don't have any custom entity manipulation, no playing with the entity map

Reply:
Ah ok got it. I think the issue was when also handling many-to-many which the ORM didn't handle well, and it was an all-or-nothing when using the entity map, so we went with not using it at all. But I think now that we are getting to a more feature-complete state, we will have time to revisit the code in the modules and see what makes most sense to do.